torch/_inductor/runtime/triton_helpers.py

```diff
@@ -384,3 +379,162 @@ def frexp(x):
     exponent = tl.where(x == 0, 0, y)
     mantissa = tl.where(x == 0, 0, libdevice.ldexp(x, -y))
     return mantissa, exponent
+
+
+@triton.jit
+def _compare_and_swap_with_index(
+    x,
+    idxs,
+    valid_mask,
+    flip,
+    i: tl.constexpr,
+    n_dims: tl.constexpr,
+    stable: tl.constexpr,
+    descending: tl.constexpr,
+):
+    n_outer: tl.constexpr = x.numel >> n_dims
+    shape: tl.constexpr = [n_outer * 2**i, 2, 2 ** (n_dims - i - 1)]
+
+    idtype = tl.core.get_int_dtype(bitwidth=x.dtype.primitive_bitwidth, signed=True)
+
+    y = tl.reshape(x, shape)
+    iy = y.to(idtype, bitcast=True)
+    # slice left/right with 'stride' 2**(n_dims - i - 1)
+    right_mask = tl.arange(0, 2)[None, :, None].to(idtype)
+    left_mask = (1 - right_mask).to(idtype)
+    ileft = tl.broadcast_to(tl.sum(iy * left_mask, 1)[:, None, :], shape)
+    iright = tl.broadcast_to(tl.sum(iy * right_mask, 1)[:, None, :], shape)
```

Comment:
We use this trick all over the place because triton's reduce does not work for non-commutative operators. Otherwise, we could do a reduce with `lambda a, b: b`. If you think it'd be beneficial (you say this kernel is not bandwidth?), I could send a PR to triton fixing this.

Reply:
Feel free to give it a shot